File: src/contextAnalyzer.ts
import { DetectedEntity, AttackType } from "./types";

export class ContextAnalyzer {
  analyzeMultiStepAttacks(
    currentPrompt: string,
    history: Array<{ prompt: string; entities: DetectedEntity[] }>
  ): { attackType: AttackType; escalationFactor: number } {
    let attackType: AttackType = "NORMAL";
    let escalationFactor = 1.0;

    if (history.length === 0) {
      return { attackType, escalationFactor };
    }

    const historyStr = history.map((h) => h.prompt.toLowerCase()).join(" ");
    const currentStr = currentPrompt.toLowerCase();

    // Pattern 1: Email disclosure followed by password request
    if (
      historyStr.includes("email") &&
      (currentStr.includes("password") || currentStr.includes("send password"))
    ) {
      attackType = "CREDENTIAL_EXFILTRATION";
      escalationFactor = 1.5;
    }

    // Pattern 2: API key in history + secret in current
    if (
      historyStr.includes("api") &&
      (currentStr.includes("secret") || currentStr.includes("key"))
    ) {
      attackType = "API_KEY_LEAK";
      escalationFactor = 1.4;
    }

    // Pattern 3: Prompt injection patterns
    if (
      currentStr.includes("ignore") ||
      currentStr.includes("forget") ||
      currentStr.includes("ignore previous") ||
      currentStr.includes("system prompt")
    ) {
      attackType = "PROMPT_INJECTION";
      escalationFactor = 1.3;
    }

    return { attackType, escalationFactor };
  }

  generateReason(
    detectedEntities: DetectedEntity[],
    keywordFindings: { keyword: string; score: number }[],
    contextAttackType: AttackType,
    escalationFactor: number
  ): string {
    const reasons: string[] = [];

    if (detectedEntities.length > 0) {
      const types = detectedEntities.map((e) => e.type).join(", ");
      reasons.push(`Detected entities: ${types}`);
    }

    if (keywordFindings.length > 0) {
      const keywords = keywordFindings.map((k) => k.keyword).join(", ");
      reasons.push(`Sensitive keywords: ${keywords}`);
    }

    if (contextAttackType !== "NORMAL") {
      reasons.push(`Multi-step attack pattern detected: ${contextAttackType}`);
    }

    if (escalationFactor > 1.0) {
      reasons.push(`Context-based escalation (${escalationFactor.toFixed(2)}x)`);
    }

    return reasons.join(" | ");
  }
}

File: context_analyzer.py
"""
Context analyzer for multi-step attack detection.
Ported from TypeScript contextAnalyzer.ts.
"""

from __future__ import annotations

import re
from typing import List

from models import AttackType, DetectedEntity, EntityType, HistoryEntry


class ContextAnalyzer:
    """Analyses prompt history to detect multi-step attack patterns."""

    INJECTION_PATTERNS: List[re.Pattern] = [
        re.compile(r"<!--[\s\S]*?-->"),              # HTML comments
        re.compile(r"/\*[\s\S]*?\*/"),                # C-style comments
        re.compile(r"<script[\s\S]*?</script>", re.I),  # Script tags
        re.compile(r"\{\{[\s\S]*?\}\}"),              # Template injection
        re.compile(r"\$\{[\s\S]*?\}"),                # Template literals
        re.compile(r"ignore\s+previous", re.I),       # Ignore previous
        re.compile(r"system\s*prompt", re.I),         # System prompt ref
        re.compile(r"\b(override|bypass)\b", re.I),   # Override/bypass
    ]

    def detect_attack_type(
        self,
        current_prompt: str,
        history: List[HistoryEntry],
    ) -> AttackType:
        # Check for prompt injection first
        if self._is_prompt_injection(current_prompt):
            return AttackType.PROMPT_INJECTION

        if not history:
            return AttackType.NORMAL

        # Check for credential exfiltration
        has_email_in_history = any(
            any(e.type == EntityType.EMAIL for e in entry.detected_entities)
            for entry in history
        )
        has_password_mention = "password" in current_prompt.lower()

        if has_email_in_history and has_password_mention:
            return AttackType.CREDENTIAL_EXFILTRATION

        # Check for API key leak pattern
        has_api_in_history = any(
            any(e.type == EntityType.API_KEY for e in entry.detected_entities)
            for entry in history
        )
        has_secret_mention = "secret" in current_prompt.lower()

        if has_api_in_history and has_secret_mention:
            return AttackType.API_KEY_LEAK

        # Check current prompt for API-related patterns
        if "api" in current_prompt.lower() and (
            "key" in current_prompt.lower() or "secret" in current_prompt.lower()
        ):
            return AttackType.API_KEY_LEAK

        return AttackType.NORMAL

    def _is_prompt_injection(self, prompt: str) -> bool:
        return any(p.search(prompt) for p in self.INJECTION_PATTERNS)

    def generate_reason(
        self,
        entities: List[DetectedEntity],
        attack_type: AttackType,
        decision: str,
    ) -> str:
        parts: List[str] = []
        if entities:
            types = ", ".join(sorted({e.type.value for e in entities}))
            parts.append(f"Detected: {types}")
        if attack_type != AttackType.NORMAL:
            parts.append(f"Attack: {attack_type.value}")
        parts.append(f"Decision: {decision}")
        return " | ".join(parts) if parts else "No sensitive data detected"
